chanfitness/chanfitness_test.go

```diff
@@ -0,0 +1,132 @@
+	"github.com/lightningnetwork/lnd/lnwire"
+)
+
+func TestMonitorChannelEvents(t *testing.T) {
+	privKey, err := btcec.NewPrivateKey(btcec.S256())
+	if err != nil {
+		t.Fatalf("Error getting pubkey: %v", err)
+	}
+
+	shortID := lnwire.ShortChannelID{
+		BlockHeight: 1234,
+		TxIndex:     2,
+		TxPosition:  2,
+	}
+
+	tests := []struct {
+		name string
+
+		cancel func()
+
+		// event queue us a set of events that should be sent into the
+		// updates channel passed to MonitorChannelEvents.
+		eventQueue []interface{}
+
+		// expectedEvents is an ordered set of events that we expect the
+		// to have been recorded by the event store for the channel.
+		expectedEvents []*channelEvent
+	}{
+
+		{
+			name:   "Open channel event",
+			cancel: func() {},
+			expectedEvents: []*channelEvent{
+				{
+					eventType: channelOpenedEvent,
+				},
+			},
+			eventQueue: []interface{}{
+				channelnotifier.OpenChannelEvent{
+					Channel: &channeldb.OpenChannel{
+						ShortChannelID: shortID,
+						IdentityPub:    privKey.PubKey(),
+					},
+				},
+			},
+		},
+		{
+			name:   "Close channel event",
+			cancel: func() {},
+			expectedEvents: []*channelEvent{
+				{
+					eventType: channelOpenedEvent,
+				},
+				{
+					eventType: channelClosedEvent,
+				},
+			},
+			eventQueue: []interface{}{
+				channelnotifier.OpenChannelEvent{
+					Channel: &channeldb.OpenChannel{
+						ShortChannelID: shortID,
+						IdentityPub:    privKey.PubKey(),
+					},
+				},
+				channelnotifier.ClosedChannelEvent{
+					CloseSummary: &channeldb.ChannelCloseSummary{
+						ShortChanID: shortID,
+						RemotePub:   privKey.PubKey(),
+					},
+				},
+			},
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			store := &ChannelEventStore{
+				channels: map[uint64]*chanEventLog{},
+				quit:     make(chan struct{}),
+			}
+
+			// Populate a buffered channel with the test event queue.
+			updates := make(chan interface{}, len(test.eventQueue))
+			for _, e := range test.eventQueue {
+				updates <- e
+			}
+
+			// Start goroutine to consume update events.
+			store.wg.Add(1)
+			go store.MonitorChannelEvents(updates, test.cancel)
+
+			// Wait until the eventlog has expected number of events to prevent
+			// data races when we kill the monitoring goroutine.
+			err := lntest.WaitPredicate(func() (ok bool) {
+				eventlog, ok := store.getEventLog(shortID.ToUint64())
+				if !ok {
+					return false
+				}
+
+				return len(test.expectedEvents) == len(eventlog.events)
+
+			}, time.Second)
+
+			if err != nil {
+				t.Fatalf("Did not get expected number of events")
+			}
+
+			// Check that the events in the log are as expected.
+			eventlog, _ := store.getEventLog(shortID.ToUint64())
+			for i, e := range test.expectedEvents {
+				if eventlog.events[i].eventType != e.eventType {
+					t.Fatalf("Expected: event type: %v, got: %v",
+						eventlog.events[i].eventType,
+						e.eventType,
+					)
+				}
+			}
+
+			store.Stop()
```

Comment:
Is this needed? Doesn't look like the store is started anywhere.

Reply:
Aah, it is to stop the goroutine. An alternative way to test this would be to start the store directly, but mock the `ChannelEvents` by changing its signature to
```go
ChannelEvents func() (<-chan interface{}, func(), error)
```

and then initiating it when creating the store:
```go
	eventStore, err := chanfitness.NewChannelEventStore(&chanfitness.Config{
		ChannelEvents:   func() (<-chan interface{}, func(), error) {
                                                      client, err := s.channelNotifier.SubscribeChannelEvents()
                                                       if err != nil {
                                                                return nil, nil, err
                                                      }
                                                      return client.Updates(), client.Cancel, nil
                                               },
		GetOpenChannels: s.chanDB.FetchAllOpenChannels,
	})
```

When starting the store in the test,  you would just return the `eventQueue` channel from this method. Would that work?

Reply:
This will get a bit messy if we're also going to listen on `client.Quit` as well, signature would have to be:
```
ChannelEvents func() (<-chan interface{}, <-chan struct{}, func(), error)
```

I guess the trade off is between this ugliness and the neatness of just being able to call `store.Start()` on tests. I'm more on the side of a little ugliness for better looking tests, what would you say?

Another option would be to run actual peer/chan notifiers for tests but that may be overkill for tests?  Would need to connect to a channel DB and all sorts.

Reply:
I see, yeah that can get messy. Can you just start a `subscribe.Server` for the test where you send the event queue to?